### PR TITLE
Port to latest hyper, Rust 2018, and async/await

### LINKED ISSUE #15: Migrate to latest hyper
Any plans on making the API asynchronous?

### PR DESCRIPTION
**Pull request by ebkalderon**
### Changed

* Port to the 2018 edition and async/await.
* Upgrade to `hyper` 0.13, replace outdated `hyper-native-tls` dependency with `hyper-tls`.
* Improve formatting and better adhere to idiomatic Rust style.
* Change `FromMap::from_map` to take `BTreeMap<String, String>`.
* Eliminate unnecessary `to_string()` allocations where possible.
* Implement `Display` and `Error` for `TwilioError`.
* Update `README.md` code snippets.

### Fixed

* Fix bug with `+` character in `url_encode()` (fix copied from PR #2).

Please test this crate on your end to make sure it works as expected! Let me know if something broke. @neil-lobracco 

Closes #15.
Closes #2.

### CODE REVIEW

File: src/call.rs
Comment (neil-lobracco) [inline src/call.rs:50]: Curious about this change - is `BTreeMap` now considered the better choice in general than `HashMap`?

**Comment by ebkalderon** [comment]
Thanks for the review, @neil-lobracco! Unfortunately, I wasn't able to test this out myself. 😞 I don't have an account that I am personally authorized to play with at the moment, but I might be able to arrange for one.

File: src/call.rs
Comment (ebkalderon) [reply src/call.rs:50]: I tried to explain the reasoning behind the type signature change in the commit message for 504df24, but the gist is, essentially as follows: the current code generates the map in either a one or two step process, depending on whether the method is a `GET` or `POST`.

If the method is a `GET`, a `Vec<(String, String)>` is created and then immediately recollected into a `HashMap<&str, &str>`. If the method is a `POST`, a `Vec<(String, String)>` is created, sorted with `Ord::cmp`, and then recollected into a `HashMap<&str, &str>`. The resulting `HashMap` is then passed by reference to `FromMap::from_map`, where nearly every single implementation then takes the contained `&str` values and calls `.to_string()` on them again, even though the original `String` values the `str` slices reference are still available. This is a lot of unnecessary work and results in quite a few redundant memory allocations.

By changing the data type to a `BTreeMap<String, String>`, passing it by-move, and eliminating the intermediate `Vec<(String, String)>`, we can eliminate some intermediate memory allocations and extra work.

1. The data structure returned by `args_from_urlencoded()` is already in a map-like form, meaning it's suitable for passing into `FromMap::from_map` and used directly. By collecting into a `BTreeMap` from the beginning, there is no need for a three-step `Vec<(String, String)> -> sort? -> HashMap<&str, &str>` conversion.
2. The map is always sorted by key according to `Ord`, meaning there's no need for an explicit `Vec::sort_by` call.
3. `FromMap::from_map` takes ownership of the map itself, and the keys and values are owned `String`s instead of `&str`, meaning the implementers of the trait can simply `remove()` owned elements from the map instead of calling `.get() -> .to_string()`.

Hope this makes some sense. Let me know if there's something I'm missing!

File: src/webhook.rs
Comment (ebkalderon) [reply src/webhook.rs:31]: D'oh! 🤦 That's a good catch. I wasn't aware that `get()` used to return an array of `HeaderValue`s instead of only one. You're correct that this would always return `TwilioError::BadRequest`, in that case. I think we can remove the `.len()` check and `match` statement, in that case. Thanks for pointing that out.

Comment (ebkalderon) [reply src/webhook.rs]: Fixed in 2e8051b.

**Comment by ebkalderon** [comment]
I definitely agree that such drastic API changes would warrant a stable 1.0 version bump, or at least a 0.x.0 version bump if you feel `twilio-rs` isn't completely ready for prime-time. I see there are other pull requests open with further improvements, some of which are, granted, subsumed or obsolesced by this PR. I wouldn't mind opening follow-up PRs incorporating these other changes into the new async codebase, if you're interested in reviewing them.

As for whether you should keep a branch open with the `0.x` code open, feel free if you think it's important! Though keep in mind, you probably won't be able to easily merge those commits into `master` after this PR is merged (unless you're willing to rebase `master` and reorder those commits before the `1.x` release, which I assume you probably wouldn't want to do). You may have to keep the `0.x` branch around forever, because any further bugfixes or improvements made to it will likely not be recorded in `master` history. Whether this is worth it is ultimately your call, I think.

File: src/call.rs
Comment (neil-lobracco) [reply src/call.rs:50]: Thanks for taking the time to explain it - makes sense to me.

**Review by neil-lobracco — APPROVED** [review]
This looks good to me - I was finally able to get to this and test sending an SMS, making a call, and receiving a webhook and things looked to be working well.
I pushed up a branch `0.x` based on current master, and I'm going to merge this and push up a new version `1.0.0` into Cargo - while I'm sure there's plenty more work that it could use, I'd imagine that most projects would prefer to be able to use `async` / `await`, as well as use modern versions of the dependencies, and we can always bump the major version if more breaking changes are needed down the road.
Thanks for putting this together! I appreciate it.

**Comment by ebkalderon** [comment]
That's wonderful! I'm glad you were able to verify the changes were working correctly. Hopefully further improvements in the other open PRs could be rebased or refactored on top of this. Thanks for creating this library, @neil-lobracco!

**Pull request merged**
